policy: Remove denied identities maps

ianvernon opened the pull request:
Now that requirement selectors are folded into the selectors of the
rules in selector policies, we no longer need to keep track of
identities denied at L3 to produce the correct L4 policies. Removal of
DeniedIngressIdentities and DeniedEgressIdentities from the
EndpointPolicy also allows the removal of the matchingRules members
from the SelectorPolicy.

The generation of a SelectorPolicy is made faster by keeping note of
the selected rules on the first pass, and then iterating only over
those rules in subsequent passes. Note that this changes the set of
rules the later passes process and may result in different policy
trace output than before.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off by: Ian Vernon <ian@cilium.io>

This is cherry-picked from https://github.com/cilium/cilium/pull/7958 to reduce the size of said PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7962)
<!-- Reviewable:end -->

joestringer commented on pkg/policy/api/egress.go:
```diff
@@ -176,6 +180,29 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
 }
 
+// GetDestinationEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
+// all L3 source selectors of the ingress rule
```
I suspect that with the rest of the other PR eliminating dead code, we might be able to fold `GetDestinationEndpointSelectorsWithRequirements()` and `GetDestinationEndpointSelectors()` together?

Also:
```suggestion
// all L3 destination selectors of the egress rule
```

jrajahalme replied:
Looked into this, and indeed, `canReach*()` functions as well as `GetDestinationEndpointSelectors()` can be removed. I'll prepare a separate commit for this in another PR.

jrajahalme replied:
Folded these dead code removals here to avoid unnecessary work fixing the dead functions.

joestringer commented on pkg/policy/repository_test.go:
```diff
@@ -1844,7 +1844,7 @@ Ingress verdict: allowed
 	expectedOut = `
 Resolving ingress policy for [any:foo]
-0/1 rules selected
+0/0 rules selected
```
There aren't 0 rules in the repository, though?

I think we just need to count the total early on, and print that here. the `0/1` is correct (no rules select 'foo').

jrajahalme replied:
So the latter number should be the total number of rules in the repository? And the first number the rules selected for "foo"?

jrajahalme replied:
Fixed :-)

joestringer commented on pkg/policy/repository.go:
```diff
@@ -143,12 +143,6 @@ func wildcardL3L4Rule(proto api.L4Proto, port int, endpoints api.EndpointSelecto
 
-// wildcardL3L4Rules updates each ingress L7 rule to allow at L7 all traffic that
-// is allowed at L3-only or L3/L4.
-func (p *Repository) wildcardL3L4Rules(ctx *SearchContext, ingress bool, l4Policy L4PolicyMap) {
```
I was worried about this at first, but on second look seems like it's just unused. :+1:

joestringer reviewed (changes requested):
Couple of minor comments, otherwise looks good!

jrajahalme committed:
policy: Remove denied identities maps

Now that requirement selectors are folded into the selectors of the
rules in selector policies, we no longer need to keep track of
identities denied at L3 to produce the correct L4 policies. Removal of
DeniedIngressIdentities and DeniedEgressIdentities from the
EndpointPolicy also allows the removal of the matchingRules members
from the SelectorPolicy.

The generation of a SelectorPolicy is made faster by keeping note of
the selected rules on the first pass, and then iterating only over
those rules in subsequent passes. Note that this changes the set of
rules the later passes process and may result in different policy
trace output than before.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme committed:
policy: Remove dead code

Remove dead code in preparation for wiring in the selector cache for
the selector policy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: fix rules count in trace output.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer reviewed (approved):
:shipit:

ianvernon reviewed (commented):
I'm approving these commits (as proxy and API owners) but I can't approve because this is my PR. But, they are Jarno's commits, so consider this approved by proxy and API codeowners :)
